File: server/src/main/scala/chess/domain/GameStatus.scala
package com.chessonline
package chess.domain

import enumeratum._

sealed abstract class GameStatus(val tag: String) extends EnumEntry

object GameStatus extends Enum[GameStatus] {
  val values: IndexedSeq[GameStatus] = findValues

  case object GameContinues extends GameStatus("gameContinues")
  final case class Win(by: Side) extends GameStatus("win")
  case object Draw extends GameStatus("draw")
}

File: client/lib/rooms/models/game_status.dart
import 'package:client/chess/models/side.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'game_status.freezed.dart';
part 'game_status.g.dart';

@Freezed(unionKey: 'tag')
class GameStatus with _$GameStatus {
  const factory GameStatus.gameContinues() = GameContinues;

  const factory GameStatus.draw() = Draw;

  const factory GameStatus.win(Side by) = Win;

  factory GameStatus.fromJson(Map<String, dynamic> json) =>
      _$GameStatusFromJson(json);
}
